### ISSUE TITLE
Allow legacy TCP connections

### ISSUE THREAD

**Issue by jakubcech** [iotaledger/iri#1378]

**Comment by luca-moser** [comment]
After further investigating the current TCP codebase, I would really prefer if we could do a hard cut where we force node operators to upgrade.

Following issues make up my opinion:
* A `TCPNeighbor` has two TCP sockets, one for reading and writing to/from the neighbor. 
*A TCP socket is bidirectional, it doesn't make any sense to have 2 sockets.*
* Sent messages are "enriched" with a CRC32 checksum, likewise, the reading end part consumes the payload to compute the CRC32 checksum and then checks it against the one in the payload. 
*On TCP, the data integrity is already taken care of by the protocol and lower level layers, this is absolutely redundant. From reading Slack, apparently someone tried to fix a race condition by introducing the CRC32 checksum. This doesn't make any sense either.*
* A 32 thread pool is allocated for receiving messages from neighbors and an additional 32 thread pool is allocated for handling sending messages to neighbors.
*A NIO selector accomplishes the same but with only 1 thread*.
* The code messages the source TCP socket port (`TCP_RECEIVER_PORT`) by encoding it to a string and then to a 10 byte array. 
*A socket port can be encoded using 2 bytes.*
* The current codebase doesn't support multiple neighbors originating from the same IP address, as the connection is dropped, if a neighbor with the same origin IP address is already connected.
*The distinguishing factor should be the origin IP address + the source TCP socket port.*

I wouldn't let anyone of my friends run this code in good faith, likewise, I wouldn't want any node operator to run this code either. We should advertise in advance that the network-rewrite will induce an incompatible network upgrade and given that most node operators are accessible through social media streams and Discord, it should not be a problem to keep up the network during the upgrade, as enough will upgrade. Clearly, I am building this on assumptions which I am happy to discuss. Obviously the board will decide what to do.

My current network rewrite has this issues all resolved but it has no knowledge on how to handle the current legacy connections.

The current "architecture" of the TCP codebase looks like this:
![legacy_tcp](https://user-images.githubusercontent.com/1385611/54697185-6ca6e800-4b2d-11e9-9432-74d1d2f06608.png)

**Comment by shufps** [comment]
omg, from my side definitely a Go for removing legacy code :+1:

**Comment by muXxer** [comment]
IMHO we should definitely remove the legacy code. Right now the network is still in beta stage, and the node operators are quite fast with updating their nodes. You can see the version fragmentation here https://iota-nodes.net/statistics. Old nodes can be dropped because their perfomance isn't worth the effort to keep the compatibility. Further improvements will force the node owners to update their IRI anyways. Otherwise I will force them with spamming to do it  :)

**Comment by jakubcech** [comment]
We should try and work introducing the breaking change as the first option. One thing to keep in mind is that it's not only node operators that are part of the community who run IRI. The community is definitely very fast to upgrade and I'm very confident that almost everyone here will be up for the breaking change 😄. But IRI is also run by exchanges, so we need to handle the transition there not to disrupt their operation. I'm not saying this can't be handled, but we'll have to reach out to them and communicate and plan the transition.

**Comment by muXxer** [comment]
It's a good idea to get the exchanges in the same boat. I think it is in their interest to have a stable and synced connection of their nodes to the rest of the network :+1:

**Comment by Citrullin** [comment]
How does it look like in cIRI? Does cIRI also need a change?

**Comment by thibault-martinez** [comment]
cIRI would also need a change. But since it's not even alpha yet, a breaking change wouldn't hurt.
Also, I am totally in favor of a breaking change if it means getting rid of this legacy TCP layer once and for all.

**Comment by muXxer** [comment]
Yes, cIRI also needs that change AFAIK. But it's definitely worth the effort. :)

**Comment by nuriel77** [comment]
This change includes making UDP peering obsolete?

**Comment by Citrullin** [comment]
@thibault-martinez Time to shut down my cIRI node :D

**Comment by GalRogozinski** [comment]
I agree that networking should be changed. I disagree with @jakubcech that there should be breaking changes introduced that rely on good communication with the community and exchanges.

We can introduce those new changes to be opt-in while still supporting the old message format (i.e. packets with CRC32). By the way inefficiencies like having two TCP sockets can be eliminated without worry since an old IRI node doesn't care how it's neighbor manages its sockets, thus it is not a breaking change.

This is the way most other respectable cryptocurrency projects do it, and this is how we should do it. We can't assume that everyone is catching up on github/discord. The figures here https://iota-nodes.net/statistics tell a partial story (only Nelson nodes are being counted) and for a beta project we hold quite a significant value.

By the way @luca-moser 
>A 32 thread pool is allocated for receiving messages from neighbors and an additional 32 thread pool is allocated for handling sending messages to neighbors.
A NIO selector accomplishes the same but with only 1 thread.

I didn't tell you but I was never entirely convinced that the NIO selector is necessarily better than the thread per connection model. See https://www.slideshare.net/e456/tyma-paulmultithreaded1. To Sum it up, NIO is more scalable since threads take up resources. It is not necessarily faster and some benchmarks found it to be slower since apparently non-contentious thread switching can be fast.
In our case a node is suppose to have a low amount of peers (3-7), and I am pretty sure that the persistence layer is currently more of a bottleneck. Both you and @akashgoswami wanted to use a selector so I simply decided to be quiet about this since I think either model can be fine. Once we reach a new networking bottleneck we can experiment again.
